File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      nothing?(input) -> "Fine. Be that way!"
      yelling_question?(input) -> "Calm down, I know what I'm doing!"
      question?(input) -> "Sure."
      all_uppercase?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp all_uppercase?(input) do
    input
    |> String.replace(~r/[\d,? ]/, "")
    |> String.match?(~r/^[^a-z]+$/u)
  end

  defp question?(input), do: input =~ ~r/[?]$/

  defp nothing?(input), do: String.trim_leading(input) == ""

  defp yelling_question?(input), do: all_uppercase?(input) and question?(input)
end

File: ruby/bob/bob.rb
class Bob
  def self.hey(text)
    text = clean_text(text)

    if text == ""
      "Fine. Be that way!"
    elsif yelling?(text) && question?(text)
      "Calm down, I know what I'm doing!"
    elsif yelling?(text)
      "Whoa, chill out!"
    elsif question?(text)
      "Sure."
    else
      "Whatever."
    end
  end

  private

  def self.clean_text(text) text.gsub(/[^a-zA-Z?!\d]/, "") end
  
  def self.question?(text) text =~ /\?+$/ end
    
  def self.yelling?(text)
    text = text.gsub(/[\d,? ]/, "")
    text =~ /^[A-Z]+[!]*$/
  end
end